File: src/corner_detector.py
import cv2
import matplotlib.pyplot as plt
from mat_to_points import mat_to_points

class CornerDetector:
    def __init__(self, margin):
        self.detector = cv2.FastFeatureDetector_create(threshold=3)
        self.keypoints = None
        self.descriptors = None
        self.kps_per_block = 1
        self.margin = margin

    def detect_keypoints(self, image, split_count):
        self._keypoints = self.detector.detect(image)
        edge = cv2.Sobel(image, cv2.CV_8U, 1, 0)

        return self._distribute_keypoints(image, edge, split_count)

    def _distribute_keypoints(self, image, edge, split_count):
        sub_width = (image.shape[1]-2*self.margin)/split_count
        sub_height = (image.shape[0]-2*self.margin)/split_count

        keypoints = []
        for i in range(0, split_count):
            for j in range(0, split_count):
                left = self.margin + int(i * sub_width)
                right = self.margin + int((i+1) * sub_width - 1)
                top = self.margin + int(j * sub_height)
                bottom = self.margin + int((j+1) * sub_height - 1)

                selection = None
                for keypoint in self._keypoints:
                    if keypoint.pt[0] < left or keypoint.pt[0] > right \
                            or keypoint.pt[1] < top or keypoint.pt[1] > bottom:
                        continue

                    if selection == None:
                        selection = keypoint
                    elif selection.response < keypoint.response:
                        selection = keypoint

                if selection == None:
                    for k in range(left, right):
                        for l in range(top, bottom):
                            if selection == None:
                                selection = cv2.KeyPoint(k, l, 1, -1, edge[l, k])
                                break
                            elif selection.response < edge[l,k]:
                                selection = cv2.KeyPoint(k, l, 1, -1, edge[l, k])
                                break

                keypoints.append(selection)

        return keypoints

File: src/lib/corner_detector.cpp
#include <vector>
#include <array>

#include <opencv2/opencv.hpp>
#include <opencv2/features2d.hpp>

#include "corner_detector.hpp"

CornerDetector::CornerDetector()
{
}

void CornerDetector::detect_keypoints(const Mat &image,
        int grid_width, int grid_height,
        vector<KeyPoint2d> &keypoints,
        vector<KeyPointInformation> &kp_info,
        int level)
{
    Ptr<FastFeatureDetector> detector = FastFeatureDetector::create();
    vector<KeyPoint> _keypoints;

    detector->detect(image, _keypoints);

    Mat edge;
    Sobel(image, edge, -1, 1, 0);

    auto top = 0;
    auto bottom = grid_height;
    while (true) {
        auto left = 0;
        auto right = grid_width;

        while (true) {
            if (right > image.cols)
                break;

            // First see if we find a FAST corner
            KeyPoint2d candidate;
            KeyPointInformation info;
            info.score = -1;
            for (auto keypoint : _keypoints) {
                if (keypoint.pt.x  < left || keypoint.pt.x >= right ||
                        keypoint.pt.y < top || keypoint.pt.y >= bottom)
                    continue;
                if (info.score < keypoint.response) {
                    info.score = keypoint.response;
                    candidate.x = keypoint.pt.x;
                    candidate.y = keypoint.pt.y;
                    info.type = KP_FAST;
                }
            }
            // If not we try to find an edgelet
            if (info.score < 0) {
                for (int k = left; k < right; k++) {
                    for (int l = top; l < bottom; l++) {
                        uint8_t response = edge.at<uint8_t>(l,k);
                        if (info.score < response) {
                            info.score = response;
                            candidate.x = k;
                            candidate.y = l;
                            info.type = KP_EDGELET;
                        }
                    }
                }
            }
            info.level = level;
            keypoints.push_back(candidate);
            kp_info.push_back(info);

            left += grid_width;
            right += grid_width;
        }

        bottom += grid_height;
        if (bottom > image.rows)
            break;
        top += grid_height;
    }
}

